lib/recursive-methods.rb

```diff
@@ -1,49 +1,131 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  # raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  raise ArgumentError if n < 0
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return s[s.length - 1] + reverse(s[0, s.length - 1])
+
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n^2)
+
+def reverse_helper(string, start = 0, last = s.length - 1)
```

Comment:
Since you're not creating a new array with each method call this is O(n) for space complexity.